### PR TITLE
Implement root element backgrounds

### LINKED ISSUE #19421: No background image on http://bholley.net/
Go to http://bholley.net/
No background image visible.
You can see background image in Firefox

### PR DESCRIPTION
**Pull request by pyfisch**
Note: #19669 is visible with background images (and some gradients).

Note 2: Maybe the root background only needs to be propagated on REPAINT?

Note 3: I am not sure what the complete implications of not sending a background color to webrender are. But setting one gives wrong results with partially transparent background colors.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #19421
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19704)

### CODE REVIEW

File: components/layout_thread/lib.rs [hunk 1]
```diff
@@ -1664,35 +1668,46 @@ impl ProfilerMetadataFactory for LayoutThread {
...
+/// Backgrounds by default are transparent. This ensures that iframes
+/// are displayed with a background from the surrounding context.
+///
+/// Spec: https://drafts.csswg.org/css-backgrounds/#body-background
+fn determine_root_background(root_flow: &mut Flow) {
+    if !root_flow.is_block_like() {
+        return;
+    }
+    let root_block = root_flow.as_mut_block();
+    let root_style = ServoArc::make_mut(&mut root_block.fragment.style);
```
Comment (emilio) [inline components/layout_thread/lib.rs]: It looks unfortunate to mutate the style from here... We should at least move the `make_mut` to where we know we are going to change it.

File: components/layout_thread/lib.rs [hunk 2]
```diff
@@ -1664,35 +1668,46 @@ impl ProfilerMetadataFactory for LayoutThread {
...
-        return transparent;
-    }
-
-    let kid_block_flow = kid.as_block();
-    kid_block_flow.fragment
-                  .style
-                  .resolve_color(kid_block_flow.fragment.style.get_background().background_color)
-                  .to_gfx_color()
+    let kid_mut_block = kid.as_mut_block();
+    let kid_background = ServoArc::make_mut(&mut kid_mut_block.fragment.style).mutate_background();
```
Comment (emilio) [inline components/layout_thread/lib.rs]: I'm pretty sure this `make_mut` is not needed, is it?

Why wouldn't just `kid_mut_block.fragment.style.get_background()` suffice?

But more importantly:

 * What guarantees that the `<body>` is a block?
 * What guarantees that it's the first block?

Not that the code before wasn't wrong, but...

Comment (pyfisch) [reply components/layout_thread/lib.rs]: > I'm pretty sure this make_mut is not needed, is it?
> Why wouldn't just kid_mut_block.fragment.style.get_background() suffice?

Well the borrow checker disagrees. In lines 1709 and 1710 `kid_background` is mutated. This prompts the compiler to tell me ``cannot borrow immutable borrowed content `*kid_background` as mutable``.

> What guarantees that the <body> is a block?

Good catch. While the `<body>` is usually a block with `body { display: inline; }` it is not. Which other [`FlowClass`](https://doc.servo.org/layout/flow/enum.FlowClass.html) besides block and inline can it be? 

> What guarantees that it's the first block?

The body element is always the first block. (Even if there is no `<body>` tag in the source code. Or the code is something like `<a href="https://mozilla.org">Mozilla</a><body>Main content</body>`.)
There must be some specifications that describes this. One part is [Optional Tags](https://html.spec.whatwg.org/multipage/syntax.html#optional-tags). Another is [HTML parsing The "after head" insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#the-after-head-insertion-mode) and [HTML parsing The "in body" insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody). Servo seems to follow all these weird rules as far as I can tell.

File: components/layout_thread/lib.rs (hunk 1 above)
Comment (pyfisch) [reply components/layout_thread/lib.rs]: > It looks unfortunate to mutate the style from here

Where should it be done instead?

File: components/layout_thread/lib.rs (hunk 2 above)
Comment (emilio) [reply components/layout_thread/lib.rs]: > The body element is always the first block. (Even if there is no <body> tag in the source code. Or the code is something like <a href="https://mozilla.org">Mozilla</a><body>Main content</body>.)

I'm pretty sure I can do `head { display: block }`.

Comment (emilio) [reply components/layout_thread/lib.rs]: Also, with script I can do all sorts of stuff.

Comment (pyfisch) [reply components/layout_thread/lib.rs]: > I'm pretty sure I can do head { display: block }.

Yes this breaks my assumption "always first block". And afaik I can't even ask flows if they are a body element.

**Comment by pyfisch** [comment]
I've pushed an updated version. It works with `head { display: block }`. But as you say it is possible to [do all sorts of stuff with script](https://gist.github.com/pyfisch/b580f9e878a732cd0ca8d0bf3096677c), so it may fail here.

**Review by emilio — COMMENTED** [review]
I think we should fix this properly. Fragments contain the `OpaqueNode`. We should be able to get the document body from `TDocument` in `layout_thread`, and stash it as an `OpaqueNode` in `LayoutContext`.

I also think that just mutating the fragment's style using `make_mut` is not great...

**Comment by pyfisch** [comment]
> Fragments contain the OpaqueNode. We should be able to get the document body from TDocument in layout_thread.

I have done this. For block-like elements I can now check if `block.fragment.node` equals the `OpaqueNode` from body. But an [`InlineFlow`](https://doc.servo.org/layout/inline/struct.InlineFlow.html) may have multiple associated fragments. What is correct here?

**Comment by emilio** [comment]
I think for inlines you need to iterate all the fragments and find the one corresponding to the body, if any. @mbrubeck can correct me here if I'm wrong.

**Comment by pyfisch** [comment]
Unfortunately the `InlineFragments` have a different `OpaqueId` from the body. Besides if the body is an inline and empty there is not even an `InlineFlow` created. If the body has multiple lines each one has an individual background gradient. 

I feel that is the wrong place to implement this. Probably it should be done in `style` or wherever the ComputedValues are assigned to elements. 

For testing:
```
<!doctype html>
<style>
    body {
        display: inline;
        background-image: linear-gradient(to right, white, black);
    }
</style>
<body>
    Hello World!<br>
    And a Happy New Year
</body>
```

**Pull request closed without merging**